crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -639,3 +655,301 @@ fn normalize_string(
+        // `"""   content   ` to `"""content`
+        let leading_whitespace = trim_end.text_len() - trim_end.trim_start().text_len();
+        let trimmed_line_range =
+            TextRange::at(offset, trim_end.text_len()).add_start(leading_whitespace);
+        if already_normalized {
+            source_text_slice(trimmed_line_range, ContainsNewlines::No).fmt(f)?;
+        } else {
+            dynamic_text(trim_end.trim_start(), Some(trimmed_line_range.start())).fmt(f)?;
+        }
+    }
+    offset += first.text_len();
+
+    // Check if we have a single line (or empty) docstring
+    if normalized[first.len()..].trim().is_empty() {
+        // * The last line is `content" """` so we need a chaperone whitespace to avoid
+        //   `content""""`.
+        // * The last line is `content\ """` so we need a chaperone whitespace to avoid
+        //   `content\"""`. Note that `content\\ """` doesn't need one while `content\\\ """` does.
+        // * For `"""\n"""` or other whitespace between the quotes, black keeps a single whitespace,
+        //   but `""""""` doesn't get one inserted.
+        let needs_space = trim_end.ends_with(string_part.preferred_quotes.style.as_char())
+            || trim_end.chars().rev().take_while(|c| *c == '\\').count() % 2 == 1
+            || (trim_end.is_empty() && !normalized.is_empty());
+        if needs_space {
+            space().fmt(f)?;
+        }
+        string_part.preferred_quotes.fmt(f)?;
+        return Ok(());
+    }
+
+    hard_line_break().fmt(f)?;
+    // We know that the normalized string has \n line endings
+    offset += "\n".text_len();
+
+    // If some line of the docstring is less indented than the function body, we pad all lines to
+    // align it with the docstring statement. Conversely, if all lines are over-indented, we strip
+    // the extra indentation. We call this stripped indentation since it's relative to the block
+    // indent printer-made indentation.
+    let stripped_indentation = lines
+        .clone()
+        // We don't want to count whitespace-only lines as miss-indented
+        .filter(|line| !line.trim().is_empty())
+        .map(count_indentation_like_black)
+        .min()
+        .unwrap_or_default();
+
+    while let Some(line) = lines.next() {
+        let is_last = lines.peek().is_none();
+        format_docstring_line(
+            line,
+            is_last,
+            offset,
+            stripped_indentation,
+            already_normalized,
+            f,
+        )?;
+        // We know that the normalized string has \n line endings
+        offset += line.text_len() + "\n".text_len();
+    }
+
+    // Same special case in the last line as for the first line
+    let trim_end = normalized.as_ref().trim_end();
+    if trim_end.ends_with(string_part.preferred_quotes.style.as_char())
+        || trim_end.chars().rev().take_while(|c| *c == '\\').count() % 2 == 1
```

Comment:
Can we extract this test into a function if it is the same (and give it a name). That would allow us to document it once.